src/app-admin.js

```diff
@@ -0,0 +1,491 @@
+import validateUser from './util/validate-user.js';
+import getAccountDb, { isAdmin } from './account-db.js';
+import config from './load-config.js';
+import bodyParser from 'body-parser';
+
+let app = express();
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(errorMiddleware);
+
+export { app as handlers };
+
+export const TOKEN_EXPIRATION_NEVER = -1;
+
+const sendErrorResponse = (res, status, reason, details) => {
+  res.status(status).send({
+    status: 'error',
+    reason,
+    details,
+  });
+};
+
+const getUserByUsername = (userName) => {
+  return (
+    getAccountDb().first('SELECT id FROM users WHERE user_name = ?', [
+      userName,
+    ]) || {}
+  );
+};
+
+const getFileById = (fileId) => {
+  return (
+    getAccountDb().first('SELECT id FROM files WHERE files.id = ?', [fileId]) ||
+    {}
+  );
+};
+
+export const getAdminSessionFromRequest = async (req, res) => {
+  let session = validateUser(req, res);
+  if (!session) {
+    return null;
+  }
+
+  if (!(await isAdmin(session.user_id))) {
+    sendErrorResponse(res, 401, 'unauthorized', 'permission-not-found');
+    return null;
+  }
+
+  return session;
+};
+
+const validateUserInput = (res, user) => {
+  if (!user.userName) {
+    sendErrorResponse(
+      res,
+      400,
+      'user-cant-be-empty',
+      'Username cannot be empty',
+    );
+    return false;
+  }
+
+  if (!user.role) {
+    sendErrorResponse(res, 400, 'role-cant-be-empty', 'Role cannot be empty');
+    return false;
+  }
+
+  const { id: roleId } =
+    getAccountDb().first('SELECT id FROM roles WHERE roles.id = ?', [
+      user.role,
+    ]) || {};
+
+  if (!roleId) {
+    sendErrorResponse(
+      res,
+      400,
+      'role-does-not-exists',
+      'Selected role does not exists',
+    );
+    return false;
+  }
+
+  return true;
+};
+
+app.get('/masterCreated/', (req, res) => {
+  const { cnt } =
+    getAccountDb().first(
+      `SELECT count(*) as cnt
+     FROM users
+     WHERE users.user_name <> '' and users.master = 1`,
+    ) || {};
+
+  res.json(cnt > 0);
+});
+
+app.get('/users/', (req, res) => {
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  const users = getAccountDb().all(
+    `SELECT users.id, user_name as userName, display_name as displayName, enabled, ifnull(master,0) as master, roles.id as role 
+     FROM users
+     JOIN user_roles ON user_roles.user_id = users.id
+     JOIN roles ON roles.id = user_roles.role_id
+     WHERE users.user_name <> ''`,
+  );
+
+  res.json(
+    users.map((u) => ({
+      ...u,
+      master: u.master === 1,
+      enabled: u.enabled === 1,
+    })),
+  );
+});
+
+app.post('/users', async (req, res) => {
+  const user = await getAdminSessionFromRequest(req, res);
+  if (!user) return;
+
+  const newUser = req.body;
+  const { id: userIdInDb } = getUserByUsername(newUser.userName);
+
+  if (!validateUserInput(res, newUser)) return;
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏same problem here: is `validateUserInput` returning some HTTP response? Is this just exiting without returning anything to the end-customer? I have no idea without digging though the implementation of `validateUserInput`

Reply:
done